src/libprojectM/projectM.cpp

```diff
@@ -454,6 +454,228 @@ void projectM::renderFrame()
 
 }
 
+
+
+
+
+
+/* This alternate rendering is as possible:
+Pipeline pipeline;
+Pipeline *comboPipeline;
+comboPipeline = renderFrameOnlyPass1(&pipeline);
+
+for each eye:
+  renderFrameOnlyPass2(comboPipeline,xoffset,yoffset,eye);
+  
+Then
+renderFrameEndOnSeparatePasses(comboPipeline); 
+for the accounting and releasing the mutex
+*/
+
+
+Pipeline * projectM::renderFrameOnlyPass1(Pipeline *pPipeline) /*pPipeline is a pointer to a Pipeline for use in pass 2. returns the pointer if it was used, else returns NULL */
+{
+#ifdef SYNC_PRESET_SWITCHES
+    pthread_mutex_lock(&preset_mutex);
+#endif
+
+#ifdef DEBUG
+    char fname[1024];
+    FILE *f = NULL;
+    int index = 0;
+    int x, y;
+#endif
+
+    timeKeeper->UpdateTimers();
+/*
+    if (timeKeeper->IsSmoothing())
+    {
+        printf("Smoothing A:%f, B:%f, S:%f\n", timeKeeper->PresetProgressA(), timeKeeper->PresetProgressB(), timeKeeper->SmoothRatio());
+    }
+    else
+    {
+        printf("          A:%f\n", timeKeeper->PresetProgressA());
+    }*/
+
+    mspf= ( int ) ( 1000.0/ ( float ) settings().fps ); //milliseconds per frame
+
+    /// @bug who is responsible for updating this now?"
+    pipelineContext().time = timeKeeper->GetRunningTime();
+    pipelineContext().presetStartTime = timeKeeper->PresetTimeA();
+    pipelineContext().frame = timeKeeper->PresetFrameA();
+    pipelineContext().progress = timeKeeper->PresetProgressA();
+
+    beatDetect->detectFromSamples();
+
+    //m_activePreset->evaluateFrame();
+
+    //if the preset isn't locked and there are more presets
+    if ( renderer->noSwitch==false && !m_presetChooser->empty() )
+    {
+        //if preset is done and we're not already switching
+        if ( timeKeeper->PresetProgressA()>=1.0 && !timeKeeper->IsSmoothing())
+        {
+            if (settings().shuffleEnabled)
+                selectRandom(false);
+            else
+                selectNext(false);
+        }
+
+        else if ((beatDetect->vol-beatDetect->vol_old>beatDetect->beat_sensitivity ) &&
+                 timeKeeper->CanHardCut())
+        {
+            // printf("Hard Cut\n");
+            if (settings().shuffleEnabled)
+                selectRandom(true);
+            else
+                selectNext(true);
+        }
+    }
+
+
+    if ( timeKeeper->IsSmoothing() && timeKeeper->SmoothRatio() <= 1.0 && !m_presetChooser->empty() )
```

Comment:
perhaps this code could be refactored? it looks copied and pasted

Reply:
It is done.